llvm/lib/Transforms/IPO/FunctionAttrs.cpp

```diff
@@ -580,6 +582,206 @@ struct ArgumentUsesTracker : public CaptureTracker {
   const SCCNodeSet &SCCNodes;
 };
 
+// A struct of argument use: a Use and the offset it accesses. This struct
+// is to track uses inside function via GEP. If GEP has a non-constant index,
+// the Offset field is nullopt.
+struct ArgumentUse {
+  Use *U;
+  std::optional<int64_t> Offset;
+};
+
+// A struct of argument access info. "Unknown" accesses are the cases like
+// unrecognized instructions, instructions that have more than one use of
+// the argument, or volatile memory accesses. "Unknown" implies "IsClobber"
+// and an empty access range.
+// Write or Read accesses can be clobbers as well for example, a Load with
+// scalable type.
+struct ArgumentAccessInfo {
+  enum class AccessType : uint8_t { Write, Read, Unknown };
+  AccessType ArgAccessType;
+  bool IsClobber = false;
```

Comment:
The "clobber" terminology here is pretty weird, because usually "clobber" means "write", while here it's more like "read".

Also, after looking at the code, I'm a bit unclear on whether we really need both AccessType::Unknown and IsClobber. Isn't one of them sufficient?

Reply:
Thanks for this comment. Cleaned up the code:
- Removed "IsClobber".
- Changed the AccessType enum to: Read/Write/WriteWithSideEffect/Unknown.
   - Read: simple load/memcpy.
   - Write: simple store/memstore/memcpy/call(write a parameter and the parameter is writeonly/readnone and nocapture).
   - WriteWithSideEffect: call instructions write a parameter but the parameter is not writeonly/readnone or nocapture.
   - Unknown: unrecognized accesses or non-simple load/store/memset/...

While inferring the `initializes` attribute in a block (from end inst to begin), we empty the `initializes` range list for WriteWithSideEffect/Unknown, union the range list of Write/WriteWithSideEffect, and subtract the range list of Read.